JCStress Test failing for `reactor.core.publisher.FluxSwitchMapStressTest.RequestAndProduceStressTest2`

```
Error:  [ERROR] reactor.core.publisher.FluxSwitchMapStressTest.RequestAndProduceStressTest2

  Results across all configurations:

  RESULT  SAMPLES     FREQ      EXPECT  DESCRIPTION
  200, 0       84   91.30%  Acceptable  Should produced exactly what was requested
  200, 1        8    8.70%  Acceptable  Should produced exactly what was requested

  Messages:
    Check test failed
    java.lang.IllegalStateException: 200, 1 [386769567455] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][sns][14][0b0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000-0b0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0010]
    [386769581151] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][ini][14][0b0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0010-0b0000_0000_0000_0000_0000_0000_0000_0001_0000_0000_0000_0000_0000_0000_0000_0000]
    [386769599615] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][sns][14][0b0000_0000_0000_0000_0000_0000_0000_0001_0000_0000_0000_0000_0000_0000_0000_0000-0b0000_0000_0000_0000_0000_0000_0000_0001_0000_0000_0000_0000_0000_0000_0000_0010]
    [386769609123] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][ini][14][0b0000_0000_0000_0000_0000_0000_0000_0001_0000_0000_0000_0000_0000_0000_0000_0010-0b0000_0000_0000_0000_0000_0000_0000_0010_0000_0000_0000_0000_0000_0000_0000_0000]
    [386769623379] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][sns][14][0b0000_0000_0000_0000_0000_0000_0000_0010_0000_0000_0000_0000_0000_0000_0000_0000-0b0000_0000_0000_0000_0000_0000_0000_0010_0000_0000_0000_0000_0000_0000_0000_0010]
    [386769632416] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][ini][14][0b0000_0000_0000_0000_0000_0000_0000_0010_0000_0000_0000_0000_0000_0000_0000_0010-0b0000_0000_0000_0000_0000_0000_0000_0011_0000_0000_0000_0000_0000_0000_0000_0000]
    [386769645841] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][sns][14][0b0000_0000_0000_0000_0000_0000_0000_0011_0000_0000_0000_0000_0000_0000_0000_0000-0b0000_0000_0000_0000_0000_0000_0000_0011_0000_0000_0000_0000_0000_0000_0000_0010]
    [386769654858] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][ini][14][0b0000_0000_0000_0000_0000_0000_0000_0011_0000_0000_0000_0000_0000_0000_0000_0010-0b0000_0000_0000_0000_0000_0000_0000_0100_0000_0000_0000_0000_0000_0000_0000_0000]
    [386769668183] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][sns][14][0b0000_0000_0000_0000_0000_0000_0000_0100_0000_0000_0000_0000_0000_0000_0000_0000-0b0000_0000_0000_0000_0000_0000_0000_0100_0000_0000_0000_0000_0000_0000_0000_0010]
    [386769682770] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][ini][14][0b0000_0000_0000_0000_0000_0000_0000_0100_0000_0000_0000_0000_0000_0000_0000_0010-0b0000_0000_0000_0000_0000_0000_0000_0101_0000_0000_0000_0000_0000_0000_0000_0000]
    [386769717375] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][sns][14][0b0000_0000_0000_0000_0000_0000_0000_0101_0000_0000_0000_0000_0000_0000_0000_0000-0b0000_0000_0000_0000_0000_0000_0000_0101_0000_0000_0000_0000_0000_0000_0000_0010]
    [386769735559] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][ini][14][0b0000_0000_0000_0000_0000_0000_0000_0101_0000_0000_0000_0000_0000_0000_0000_0010-0b0000_0000_0000_0000_0000_0000_0000_0110_0000_0000_0000_0000_0000_0000_0000_0000]
    [386769760074] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][sns][14][0b0000_0000_0000_0000_0000_0000_0000_0110_0000_0000_0000_0000_0000_0000_0000_0000-0b0000_0000_0000_0000_0000_0000_0000_0110_0000_0000_0000_0000_0000_0000_0000_0010]
    [386769776976] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][ini][14][0b0000_0000_0000_0000_0000_0000_0000_0110_0000_0000_0000_0000_0000_0000_0000_0010-0b0000_0000_0000_0000_0000_0000_0000_0111_0000_0000_0000_0000_0000_0000_0000_0000]
    [386769853539] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][sns][14][0b0000_0000_0000_0000_0000_0000_0000_0111_0000_0000_0000_0000_0000_0000_0000_0000-0b0000_0000_0000_0000_0000_0000_0000_0111_0000_0000_0000_0000_0000_0000_0000_0010]
    [386769902600] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][ini][14][0b0000_0000_0000_0000_0000_0000_0000_0111_0000_0000_0000_0000_0000_0000_0000_0010-0b0000_0000_0000_0000_0000_0000_0000_1000_0000_0000_0000_0000_0000_0000_0000_0000]
    [386769935071] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][sns][14][0b0000_0000_0000_0000_0000_0000_0000_1000_0000_0000_0000_0000_0000_0000_0000_0000-0b0000_0000_0000_0000_0000_0000_0000_1000_0000_0000_0000_0000_0000_0000_0000_0010]
    [386770000549] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][adr][16][0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_0000_0000_0000-0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_0000_0001_0000]
    [386770003268] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][ini][14][0b0000_0000_0000_0000_0000_0000_0000_1000_0000_0000_0000_0000_0000_0000_0000_0010-0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_0000_0000_0000]
    [386770019835] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][adr][16][0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_0000_0001_0000-0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_0000_0010_0000]
    [386770028164] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][sns][14][0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_0000_0010_0000-0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_0000_0010_0010]
    [386770036076] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][adr][16][0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_0000_0010_0010-0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_0000_0011_0010]
    [386770047390] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][swp][14][0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_0000_0011_0010-0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_0000_0011_0011]
    [386770051665] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][adr][16][0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_0000_0011_0011-0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_0000_0100_0011]
    [386770068456] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][adr][16][0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_0000_0100_0011-0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_0000_0101_0011]
    [386774687609] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][swp][14][0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_1100_1000_0010-0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_1100_1000_0011]
    [386774694953] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][sic][14][0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_1100_1000_0011-0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_1100_1000_0110]
    [386774702177] [reactor.core.publisher.FluxSwitchMapNoPrefetch$SwitchMapMain@85655fe][smc][14][0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_1100_1000_0110-0b0000_0000_0000_0000_0000_0000_0000_1001_0000_0000_0000_0000_0000_1100_1000_1110]
    	at reactor.core.publisher.FluxSwitchMapStressTest$RequestAndProduceStressTest2.arbiter(FluxSwitchMapStressTest.java:319)
    	at reactor.core.publisher.FluxSwitchMapStressTest_RequestAndProduceStressTest2_jcstress$TestResourceEstimator.runWith(FluxSwitchMapStressTest_RequestAndProduceStressTest2_jcstress.java:266)
    	at org.openjdk.jcstress.infra.runners.ForkedTestConfig.tryWith(ForkedTestConfig.java:126)
    	at org.openjdk.jcstress.infra.runners.ForkedTestConfig.adjustStrideCount(ForkedTestConfig.java:97)
    	at reactor.core.publisher.FluxSwitchMapStressTest_RequestAndProduceStressTest2_jcstress.jcstress_sanityCheck_Resource(FluxSwitchMapStressTest_RequestAndProduceStressTest2_jcstress.java:279)
    	at reactor.core.publisher.FluxSwitchMapStressTest_RequestAndProduceStressTest2_jcstress.sanityCheck(FluxSwitchMapStressTest_RequestAndProduceStressTest2_jcstress.java:30)
    	at org.openjdk.jcstress.infra.runners.Runner.run(Runner.java:64)
    	at org.openjdk.jcstress.ForkedMain.main(ForkedMain.java:79)
```

--- a/reactor-core/src/jcstress/java/reactor/core/publisher/FluxSwitchMapStressTest.java
+++ b/reactor-core/src/jcstress/java/reactor/core/publisher/FluxSwitchMapStressTest.java
@@ -272,7 +272,8 @@ public abstract class FluxSwitchMapStressTest {
 
 	// Ignore, flaky test (https://github.com/reactor/reactor-core/issues/3633)
 	//@JCStressTest
-	@Outcome(id = {"200, 0", "200, 1"}, expect = ACCEPTABLE, desc = "Should produced exactly what was requested")
+	@Outcome(id = {"200, 0, 0", "200, 0, 1"}, expect = ACCEPTABLE, desc = "Should " +
+			"produced exactly what was requested")
 	@State
 	public static class RequestAndProduceStressTest2 extends FluxSwitchMapStressTest {
 
@@ -312,9 +313,17 @@ public abstract class FluxSwitchMapStressTest {
 		}
 
 		@Arbiter
-		public void arbiter(II_Result r) {
-			r.r1 = (int) (stressSubscriber.onNextCalls.get() + switchMapMain.requested);
-			r.r2 = stressSubscriber.onCompleteCalls.get();
+		public void arbiter(III_Result r) {
+			r.r1 = stressSubscriber.onNextCalls.get();
+			r.r2 = (int) switchMapMain.requested;
+			r.r3 = stressSubscriber.onCompleteCalls.get();
+
+			switch (r.toString()) {
+				case "200, 0, 0":
+				case "200, 0, 1":
+					break;
+				default: throw new IllegalStateException(r + " " + fastLogger);
+			}
 
 			if (stressSubscriber.onNextCalls.get() < 200 && stressSubscriber.onNextDiscarded.get() < switchMapMain.requested) {
 				throw new IllegalStateException(r + " " + fastLogger);

--- a/reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java
+++ b/reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java
@@ -103,6 +103,7 @@ final class ColdTestPublisher<T> extends TestPublisher<T> {
 		}
 
 		s.onSubscribe(p); // will trigger drain() via request()
+		p.drain(); // ensures that empty source terminal signal is propagated without waiting for a request from the subscriber
 	}
 
 	boolean add(ColdTestPublisherSubscription<T> s) {
@@ -315,7 +316,7 @@ final class ColdTestPublisher<T> extends TestPublisher<T> {
 		 * @return true if the TestPublisher was terminated, false otherwise
 		 */
 		private boolean emitTerminalSignalIfAny() {
-			if (parent.done) {
+			if (parent.done && this.parent.values.size() == index) {
 				parent.remove(this);
 
 				final Throwable t = parent.error;

--- a/reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java
+++ b/reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017-2021 VMware Inc. or its affiliates, All Rights Reserved.
+ * Copyright (c) 2017-2024 VMware Inc. or its affiliates, All Rights Reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.

